packages/webdriverio/src/utils.js

```diff
@@ -9,135 +9,182 @@ import isObject from 'lodash.isobject'
+        selector.indexOf('*/') === 0) {
+        return 'xpath'
+    }
+    // use link text strategy if selector starts with =
+    if (selector.indexOf('=') === 0) {
+        return 'link text'
+    }
+    // use partial link text strategy if selector starts with *=
+    if (selector.indexOf('*=') === 0) {
+        return 'partial link text'
+    }
+    // recursive element search using the UiAutomator library (Android only)
+    if (selector.indexOf('android=') === 0) {
+        return '-android uiautomator'
+    }
+    // recursive element search using the UIAutomation library (iOS-only)
+    if (selector.indexOf('ios=') === 0) {
+        return '-ios uiautomation'
+    }
+    // recursive element search using accessibility id
+    if (selector.indexOf('~') === 0) {
+        return 'accessibility id'
+    }
+    // class name mobile selector
+    // for iOS = UIA...
+    // for Android = android.widget
+    if (selector.slice(0, 3) === 'UIA' || selector.slice(0, 15) === 'XCUIElementType' || selector.slice(0, 14).toLowerCase() === 'android.widget') {
+        return 'class name'
+    }
+    // use tag name strategy if selector contains a tag
+    // e.g. "<div>" or "<div />"
+    if (selector.search(/<[a-zA-Z-]+( \/)*>/g) >= 0) {
+        return 'tag name'
+    }
+    // use name strategy if selector queries elements with name attributes for JSONWP
+    // or if isMobile is used even when w3c is used
+    // e.g. "[name='myName']" or '[name="myName"]'
+    if (selector.search(/^\[name=("|')([a-zA-z0-9\-_. ]+)("|')]$/) >= 0) {
+        return 'name'
+    }
+    // allow to move up to the parent or select current element
+    if (selector === '..' || selector === '.') {
+        return 'xpath'
+    }
+    // any element with given class, id, or attribute and content
+    // e.g. h1.header=Welcome or [data-name=table-row]=Item or #content*=Intro
+    if (selector.match(new RegExp([
+        // HTML tag
+        /^([a-z0-9|-]*)/,
+        // optional . or # + class or id
+        /(?:(\.|#)(-?[_a-zA-Z]+[_a-zA-Z0-9-]*))?/,
+        // optional [attribute-name="attribute-selector"]
+        /(?:\[(-?[_a-zA-Z]+[_a-zA-Z0-9-]*)(?:=(?:"|')([a-zA-z0-9\-_. ]+)(?:"|'))?\])?/,
+        // *=query or =query
+        /(\*)?=(.+)$/,
+    ].map(rx => rx.source).join('')))) {
+        return 'xpath extended'
+    }
+}
 
+export const findStrategy = function (selector, isW3C, isMobile) {
     /**
-     * check if user has specified locator strategy directly
+     * set default strategy
      */
-    const match = value.match(DIRECT_SELECTOR_REGEXP)
-    if (match) {
-        /**
-         * ensure selector strategy is supported
-         */
+    let using = DEFAULT_STRATEGY
+    let value = selector
+
+    switch (defineStrategy(selector)) {
+    // user has specified locator strategy directly
+    case 'directly': {
+        const match = selector.match(DIRECT_SELECTOR_REGEXP)
         if (!isMobile && isW3C && !W3C_SELECTOR_STRATEGIES.includes(match[1])) {
```

Comment:
can we refactor code so that this method is condition used only once or move it to method?

Reply:
here I did not find a way to do it. We need condition in `defineStrategy` method, but we also need `match` value in `findStrategy` method. I think it is good enough. If you see any chance to reduce it I'm happy to tackle it